File: 2024/11/adventOfCode11-2.py
#
# Advent of code 11-1
#
import os
import sys
from fileinput import isfirstline
from os.path import split

script_folder_path = os.path.dirname(os.path.abspath(sys.argv[0]))
lineFilePath = script_folder_path + "\\data.txt"
fileContent = open(lineFilePath).read()

rocksLine = fileContent

results = {'0':'1'}

def addToKeyValueArray(array, key, value):
    alreadyExist = False
    for i, arrayValues in array.items():
        if i == key:
            alreadyExist = True
            break
    if not alreadyExist:
        array.setdefault(key,value)
    # print(array)
    return array

def numberOfRock(_rocks, _blink, _results):
    _blinkCounter = 0
    while _blinkCounter <= _blink:
        # print("------------------------------")
        # print(_rocks)
        _tempRocks = ""
        for _rock in _rocks.split(" "):
            _rockDone = False
            for i, _result in _results.items():
                if i == _rock:
                    _tempRocks+=" " + _result
                    _rockDone = True
            if len(_rock)% 2 == 0 and not _rockDone:
                _newNumer = " ".join([str(int(_rock[:len(_rock) // 2])), str(int(_rock[len(_rock) // 2:]))])
                _tempRocks+=" " + _newNumer
                _results = addToKeyValueArray(_results, _rock,_newNumer)
                _rockDone = True
            elif not _rockDone:
                # print(_rock)
                _newNumer = str(int(_rock) * 2024)
                _tempRocks += " " + _newNumer
                _results = addToKeyValueArray(_results, _rock,_newNumer)
        # print(str(_blinkCounter) + "/" + str(_blink))
        _blinkCounter+=1
        _rocks = _tempRocks.lstrip()
        # print(_results)
    return _results, len(_rocks.split(" "))

tmpRocksLine = ""

blinkNeeded = 25
blinkCounter = 1
nbRocks=0
counter = 1
print("Starting...")

for singleRock in rocksLine.split(" "):
    results, nbRocksForRun = numberOfRock(singleRock, blinkNeeded, results)
    nbRocks+=nbRocksForRun
    counter+=1
    print(str(counter) + "/" + str(len(rocksLine.split(" "))))

# print(results)

print("Number of blink: " + str(blinkNeeded))
print("number of rocks: " + str(nbRocks))
print("expected resut:  183435")
# < 192366
# 183435

File: 2024/11/adventOfCode11-2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_ROCKS 1000
#define MAX_LENGTH 100

char* addToKeyValueArray(char results[][2][MAX_LENGTH], int *size, const char *key, const char *value) {
    for (int i = 0; i < *size; i++) {
        if (strcmp(results[i][0], key) == 0) {
            return results[i][1];
        }
    }
    strcpy(results[*size][0], key);
    strcpy(results[*size][1], value);
    (*size)++;
    return results[*size - 1][1];
}

void numberOfRock(const char *rocks, int blink, char results[][2][MAX_LENGTH], int *size, int *rockCount) {
    int blinkCounter = 0;
    while (blinkCounter <= blink) {
        char tempRocks[MAX_LENGTH] = "";
        char *rock = strtok((char *)rocks, " ");
        while (rock != NULL) {
            int rockDone = 0;
            for (int i = 0; i < *size; i++) {
                if (strcmp(results[i][0], rock) == 0) {
                    strcat(tempRocks, " ");
                    strcat(tempRocks, results[i][1]);
                    rockDone = 1;
                    break;
                }
            }
            if (strlen(rock) % 2 == 0 && !rockDone) {
                char newNumer[MAX_LENGTH];
                snprintf(newNumer, sizeof(newNumer), "%d %d", atoi(rock) / 2, atoi(rock) / 2);
                strcat(tempRocks, " ");
                strcat(tempRocks, newNumer);
                addToKeyValueArray(results, size, rock, newNumer);
                rockDone = 1;
            } else if (!rockDone) {
                char newNumer[MAX_LENGTH];
                snprintf(newNumer, sizeof(newNumer), "%d", atoi(rock) * 2024);
                strcat(tempRocks, " ");
                strcat(tempRocks, newNumer);
                addToKeyValueArray(results, size, rock, newNumer);
            }
            rock = strtok(NULL, " ");
        }
        blinkCounter++;
        strcpy((char *)rocks, tempRocks + 1); // Skip the leading space
    }
    *rockCount = strlen((char *)rocks) > 0 ? (int)(strlen((char *)rocks) - strlen(strchr((char *)rocks, ' '))) : 0;
}

int main() {
    char lineFilePath[MAX_LENGTH];
    char rocksLine[MAX_LENGTH] = "1 2 3 4 5"; // Placeholder for file content
    char results[MAX_ROCKS][2][MAX_LENGTH] = {{"0", "1"}};
    int size = 1;

    int blinkNeeded = 25;
    int blinkCounter = 1;
    int nbRocks = 0;
    int counter = 1;

    printf("Starting...\n");

    char *singleRock = strtok(rocksLine, " ");
    while (singleRock != NULL) {
        int nbRocksForRun = 0;
        numberOfRock(singleRock, blinkNeeded, results, &size, &nbRocksForRun);
        nbRocks += nbRocksForRun;
        counter++;
        printf("%d/%d\n", counter, (int)(strlen(rocksLine) - strlen(strchr(rocksLine, ' '))));
        singleRock = strtok(NULL, " ");
    }

    printf("Number of blink: %d\n", blinkNeeded);
    printf("number of rocks: %d\n", nbRocks);
    printf("expected result:  183435\n");

    return 0;
}

